### ISSUE TITLE
Add an option so `minify` can return a promise

### ISSUE THREAD

**Issue by jakubpawlowicz** [jakubpawlowicz/clean-css#856]
See https://github.com/jakubpawlowicz/clean-css/issues/842#issuecomment-267652497

**Comment by alexlamsl** [comment]
@jakubpawlowicz web usage aside, do we know the usage statistics of synchronous `new CleanCSS().minify()` on Node.js?

I ask because `html-minifier` only has a synchronous API currently, and the use of Promise would conflict with that. Obviously I can just change everything to asynchronous and publish 4.0 as well 😎

**Comment by jakubpawlowicz** [comment]
I have no idea what's the usage or where to get it.

You'd need to use sth like:

```js
var minified = new CleanCSS(options)
  .minify(input)
  .then(function (output) { return output.styles; });
```

instead of

```js
var minified = new CleanCSS(options).minify(input).styles;
```

right?

I'm not sold to introducing Promise API yet as 4.0 already has a lot of changes.

**Comment by alexlamsl** [comment]
@jakubpawlowicz in your example above, `minified` would be the (chained) Promise object as opposed to the desired output. So I can do `minified.then(function(styles){...})`, but it would still be asynchronous.

**Comment by jakubpawlowicz** [comment]
:man_facepalming: yeah, you are right, it's not that straightforward to switch to promises.

**Comment by scniro** [comment]
If helpful at all, I could envision an api change to the following. Here is the current implementation...
```
new CleanCSS().minify(source, function (error, minified) {
    //[...]
});
```

With native promises, this could be crafted as such...

```
CleanCSS.minify().then(function(minified) {
  // success handler
}, function(error) {
  // error handler
});
```

I like having the dedicated success and error handler, as oppoesd to differing return parameters. This becomes more interesting if you consider that `CleanCSS.minify` returns a promise. Imagine having the ability to have a few tasks queued up all that are promises. We can then leverage promise resolution flows in ways that are appealing. One possible example could be [`Promise.all`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all). Imagine this scenario...

```
var p1 = CleanCSS(opt);
var p2 = SomeOtherLibrary();

Promise.all([p1, p2]).then(values => { 
  console.log(values); // [minified, otherUsefulStuff] 
});
```

Other libraries such as [async](http://caolan.github.io/async/) offer some powerful flow solutions to all things "then-able." It's becoming more and more accepted that libraries are promised driven than using traditional callbacks to take advantage of these resolvable approaches that we've all beaten our minds up for when we eventually end up in callback hell.

I support either way and think the new changes coming are great - I am simply chiming here with a humble opinion on the matter 😸

**Comment by jakubpawlowicz** [comment]
Thanks @scniro, that's a very good point indeed!

@alexlamsl suggested we can use a flag saying if a promise should be returned or not (defaulting to `false`). I actually like the idea as we could 1) test how it works 2) allow both interfaces at the same time 3) flip the switch at some point if needed

**Comment by jakubpawlowicz** [comment]
@alexlamsl @scniro @ben-eb 23637f627b7ab1654a17fd6fe2633179e8cf89a1 should be it

**Comment by scniro** [comment]
@jakubpawlowicz you baller 💵

**Comment by jakubpawlowicz** [comment]
Cool, it's merged to master and due in 4.0.
